apps/src/lib/proto/types.rs

```diff
@@ -0,0 +1,329 @@
+use std::collections::hash_map::DefaultHasher;
+use std::hash::{Hash, Hasher};
+
+use prost::Message;
+use thiserror::Error;
+
+use super::generated::{services, types};
+
+#[derive(Error, Debug)]
+pub enum Error {
+    #[error("Error decoding a transaction from bytes: {0}")]
+    TxDecodingError(prost::DecodeError),
+    #[error("Error decoding an IntentGossipMessage from bytes: {0}")]
+    IntentDecodingError(prost::DecodeError),
+    #[error("Error decoding an DkgGossipMessage from bytes: {0}")]
+    DkgDecodingError(prost::DecodeError),
+    #[error("Intent is empty")]
+    NoIntentError,
+    #[error("Dkg is empty")]
+    NoDkgError,
+}
+
+pub type Result<T> = std::result::Result<T, Error>;
+
+#[derive(Debug)]
+pub struct Tx {
+    inner: types::Tx,
+}
+
+impl Tx {
+    pub fn from(tx_bytes: impl AsRef<[u8]>) -> Result<Self> {
+        let inner = types::Tx::decode(tx_bytes.as_ref())
+            .map_err(Error::TxDecodingError)?;
+        Ok(Tx { inner })
+    }
+
+    pub fn new(code: Vec<u8>, data: Option<Vec<u8>>) -> Self {
+        let inner = types::Tx {
+            code,
+            data,
+            timestamp: Some(std::time::SystemTime::now().into()),
+        };
+        Tx { inner }
+    }
+
+    pub fn to_bytes(&self) -> Vec<u8> {
+        let mut bytes = vec![];
+        self.inner
+            .encode(&mut bytes)
+            .expect("encoding a transaction failed");
+        bytes
+    }
```

Comment:
Same idea for these one :
```rust
impl ToBytes for Tx {
     fn to_bytes(self) -> Bytes {
        let mut bytes = vec![];
        self.inner
            .encode(&mut bytes)
            .expect("encoding a transaction failed");
        bytes
        }
}
```

Reply:
I don't think there is such a trait?

Reply:
oh, you're right It's part of the `bytes` crates that we don't use. I looked online a bit too quickly